RuminantFarmSystems/RuFaS
# Disentangle IM initialization

This PR addresses the issue of unintended initialization of IM prior to the official start of the program using `task_manager.start`.
### Context

Issue(s) closed by this pull request: closes #1747 

      
How big are the changes in the PR? Would you nominate it for a major version upgrade?

### What


- Most global initialization of the IM has now been moved to the constructor or within the relevant functions, ensuring that importing these packages does not automatically initiate the IM.
- The unused import and initialization of IM in `cows.py` have been removed.
- IM will be initiated as an instance variable when most of the use cases are not within static methods.

**Note:** Due to the complexity of how IM is used and to make it more testable for now, herd factory inits a new IM in every function.

### Why

Ideally, Input Manager should not be instantiated before TaskManager starts and instantiates Input Manager in TM.start(). It should also not get unexpectedly instantiated through package/class imports.

### How

This PR delays the initialization of IM, so now imports will not initialize IM until it's needed and being used after starting the tasks.

### Test plan


1. import traceback to `input_manager.py`.
2. Add this code to `input_manager.py`'s constructor: 
```
def __init__(self, metadata_depth_limit: int | None = None) -> None:
        if InputManager.__instance is None:
            print("Call stack for InputManager instantiation:")
            traceback.print_stack()
```
3. Trace and make sure that the first call is from the line calling `task_manager.start`
